moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -139,7 +144,74 @@ bool RuckigSmoothing::applySmoothing(robot_trajectory::RobotTrajectory& trajecto
     }
   }
 
-  return runRuckig(trajectory, ruckig_input);
+  auto ruckig_result = runRuckigInBatches(num_waypoints, trajectory, ruckig_input);
+  if (ruckig_result.has_value())
+  {
+    trajectory = ruckig_result.value();
+  }
+  return ruckig_result.has_value();  // Ruckig failed to smooth the trajectory
+}
+
+std::optional<robot_trajectory::RobotTrajectory>
+RuckigSmoothing::runRuckigInBatches(const size_t num_waypoints, const robot_trajectory::RobotTrajectory& trajectory,
+                                    ruckig::InputParameter<ruckig::DynamicDOFs>& ruckig_input)
+{
+  // We take the batch size as the lesser of 0.1*num_waypoints or 100, to keep a balance between run time and
+  // time-optimality.
+  // TODO(andyz): parameterize as MIN_BATCH_SIZE and BATCH_SCALING_FACTOR or something like that
+  const size_t waypoint_batch_size = [num_waypoints]() {
+    const size_t temp_batch_size = std::min(size_t(0.1 * num_waypoints), size_t(100));
+    // We need at least 2 waypoints
+    return std::max(size_t(2), temp_batch_size);
+  }();
+  size_t batch_start_idx = 0;
+  size_t batch_end_idx = waypoint_batch_size - 1;
+  const size_t full_traj_final_idx = num_waypoints - 1;
+  // A deep copy is not needed since the waypoints are cleared immediately
+  robot_trajectory::RobotTrajectory sub_trajectory =
+      robot_trajectory::RobotTrajectory(trajectory, false /* deep copy */);
+  robot_trajectory::RobotTrajectory output_trajectory =
+      robot_trajectory::RobotTrajectory(trajectory, false /* deep copy */);
+  output_trajectory.clear();
+
+  while (batch_end_idx <= full_traj_final_idx)
+  {
+    sub_trajectory.clear();
+    for (size_t waypoint_idx = batch_start_idx; waypoint_idx <= batch_end_idx; ++waypoint_idx)
+    {
+      sub_trajectory.addSuffixWayPoint(trajectory.getWayPoint(waypoint_idx),
+                                       trajectory.getWayPointDurationFromPrevious(waypoint_idx));
+    }
+
+    // When starting a new batch, set the last Ruckig output equal to the new, starting robot state
+    bool first_point_previously_smoothed = false;
+    if (output_trajectory.getWayPointCount() > 0)
+    {
+      sub_trajectory.addPrefixWayPoint(output_trajectory.getLastWayPoint(), 0);
+      first_point_previously_smoothed = true;
+    }
+
+    if (!runRuckig(sub_trajectory, ruckig_input))
+    {
+      return std::nullopt;
+    }
+
+    // Skip appending the first waypoint in sub_trajectory if it was smoothed in
+    // the previous iteration
+    size_t first_new_waypoint = first_point_previously_smoothed ? 1 : 0;
+
+    // Add smoothed waypoints to the output
+    for (size_t waypoint_idx = first_new_waypoint; waypoint_idx < sub_trajectory.getWayPointCount(); ++waypoint_idx)
```

Comment:
Good idea but I don't think that will work because of the `dt` argument. We might not have a constant `dt`.

```
  RobotTrajectory& append(const RobotTrajectory& source, double dt, size_t start_index = 0,
                          size_t end_index = std::numeric_limits<std::size_t>::max());
```

Reply:
it's only the dt between the two trajectories. All other steps remain the same

Reply:
:+1:  I replaced the second loop. The first loop -->  `[move_group-4] [WARN] [1671207429.602316971] [RobotTrajectory]: First and only waypoint has a duration of 0.`

Reply:
Sigh, I think there may be a bug in `append()`

Reply:
Bug report here:  https://github.com/ros-planning/moveit2/pull/1813

Reply:
Can we merge this in the meantime?